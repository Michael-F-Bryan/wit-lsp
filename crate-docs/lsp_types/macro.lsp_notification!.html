<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.lsp_notification.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.lsp_notification.html">macro.lsp_notification.html</a>...</p>
    <script>location.replace("macro.lsp_notification.html" + location.search + location.hash);</script>
</body>
</html>