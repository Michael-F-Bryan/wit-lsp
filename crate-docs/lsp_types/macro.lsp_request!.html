<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.lsp_request.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.lsp_request.html">macro.lsp_request.html</a>...</p>
    <script>location.replace("macro.lsp_request.html" + location.search + location.hash);</script>
</body>
</html>