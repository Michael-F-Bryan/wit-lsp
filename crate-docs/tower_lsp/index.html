<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Language Server Protocol (LSP) server abstraction for Tower."><title>tower_lsp - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-5bc39a1768837dd0.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="tower_lsp" data-themes="" data-resource-suffix="" data-rustdoc-version="1.77.2 (25ef9e3d8 2024-04-09)" data-channel="1.77.2" data-search-js="search-dd67cee4cfa65049.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-4c98445ec4002617.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-48f368f3872407c8.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-04d5337699b92874.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../tower_lsp/index.html">tower_lsp</a><span class="version">0.20.0</span></h2></div><div class="sidebar-elems"><ul class="block">
            <li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li><li><a href="#attributes">Attribute Macros</a></li></ul></section></div></nav><div class="sidebar-resizer"></div>
    <main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../tower_lsp/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">tower_lsp</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../src/tower_lsp/lib.rs.html#1-1336">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Language Server Protocol (LSP) server abstraction for <a href="https://github.com/tower-rs/tower">Tower</a>.</p>
<h2 id="example"><a class="doc-anchor" href="#example">§</a>Example</h2>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>tower_lsp::jsonrpc::Result;
<span class="kw">use </span>tower_lsp::lsp_types::<span class="kw-2">*</span>;
<span class="kw">use </span>tower_lsp::{Client, LanguageServer, LspService, Server};

<span class="attr">#[derive(Debug)]
</span><span class="kw">struct </span>Backend {
    client: Client,
}

<span class="attr">#[tower_lsp::async_trait]
</span><span class="kw">impl </span>LanguageServer <span class="kw">for </span>Backend {
    <span class="kw">async fn </span>initialize(<span class="kw-2">&amp;</span><span class="self">self</span>, <span class="kw">_</span>: InitializeParams) -&gt; <span class="prelude-ty">Result</span>&lt;InitializeResult&gt; {
        <span class="prelude-val">Ok</span>(InitializeResult {
            capabilities: ServerCapabilities {
                hover_provider: <span class="prelude-val">Some</span>(HoverProviderCapability::Simple(<span class="bool-val">true</span>)),
                completion_provider: <span class="prelude-val">Some</span>(CompletionOptions::default()),
                ..Default::default()
            },
            ..Default::default()
        })
    }

    <span class="kw">async fn </span>initialized(<span class="kw-2">&amp;</span><span class="self">self</span>, <span class="kw">_</span>: InitializedParams) {
        <span class="self">self</span>.client
            .log_message(MessageType::INFO, <span class="string">"server initialized!"</span>)
            .<span class="kw">await</span>;
    }

    <span class="kw">async fn </span>shutdown(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; <span class="prelude-ty">Result</span>&lt;()&gt; {
        <span class="prelude-val">Ok</span>(())
    }

    <span class="kw">async fn </span>completion(<span class="kw-2">&amp;</span><span class="self">self</span>, <span class="kw">_</span>: CompletionParams) -&gt; <span class="prelude-ty">Result</span>&lt;<span class="prelude-ty">Option</span>&lt;CompletionResponse&gt;&gt; {
        <span class="prelude-val">Ok</span>(<span class="prelude-val">Some</span>(CompletionResponse::Array(<span class="macro">vec!</span>[
            CompletionItem::new_simple(<span class="string">"Hello"</span>.to_string(), <span class="string">"Some detail"</span>.to_string()),
            CompletionItem::new_simple(<span class="string">"Bye"</span>.to_string(), <span class="string">"More detail"</span>.to_string())
        ])))
    }

    <span class="kw">async fn </span>hover(<span class="kw-2">&amp;</span><span class="self">self</span>, <span class="kw">_</span>: HoverParams) -&gt; <span class="prelude-ty">Result</span>&lt;<span class="prelude-ty">Option</span>&lt;Hover&gt;&gt; {
        <span class="prelude-val">Ok</span>(<span class="prelude-val">Some</span>(Hover {
            contents: HoverContents::Scalar(
                MarkedString::String(<span class="string">"You're hovering!"</span>.to_string())
            ),
            range: <span class="prelude-val">None
        </span>}))
    }
}

<span class="attr">#[tokio::main]
</span><span class="kw">async fn </span>main() {
    <span class="kw">let </span>stdin = tokio::io::stdin();
    <span class="kw">let </span>stdout = tokio::io::stdout();

    <span class="kw">let </span>(service, socket) = LspService::new(|client| Backend { client });
    Server::new(stdin, stdout, socket).serve(service).<span class="kw">await</span>;
}</code></pre></div>
</div></details><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><code>pub extern crate <a class="mod" href="../lsp_types/index.html" title="mod lsp_types">lsp_types</a>;</code></div></li></ul><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="jsonrpc/index.html" title="mod tower_lsp::jsonrpc">jsonrpc</a></div><div class="desc docblock-short">A subset of JSON-RPC types used by the Language Server Protocol.</div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.Client.html" title="struct tower_lsp::Client">Client</a></div><div class="desc docblock-short">Handle for communicating with the language client.</div></li><li><div class="item-name"><a class="struct" href="struct.ClientSocket.html" title="struct tower_lsp::ClientSocket">ClientSocket</a></div><div class="desc docblock-short">A loopback channel for server-to-client communication.</div></li><li><div class="item-name"><a class="struct" href="struct.ExitedError.html" title="struct tower_lsp::ExitedError">ExitedError</a></div><div class="desc docblock-short">Error that occurs when attempting to call the language server after it has already exited.</div></li><li><div class="item-name"><a class="struct" href="struct.LspService.html" title="struct tower_lsp::LspService">LspService</a></div><div class="desc docblock-short">Service abstraction for the Language Server Protocol.</div></li><li><div class="item-name"><a class="struct" href="struct.LspServiceBuilder.html" title="struct tower_lsp::LspServiceBuilder">LspServiceBuilder</a></div><div class="desc docblock-short">A builder to customize the properties of an <code>LspService</code>.</div></li><li><div class="item-name"><a class="struct" href="struct.Server.html" title="struct tower_lsp::Server">Server</a></div><div class="desc docblock-short">Server for processing requests and responses on standard I/O or TCP.</div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.LanguageServer.html" title="trait tower_lsp::LanguageServer">LanguageServer</a></div><div class="desc docblock-short">Trait implemented by language server backends.</div></li><li><div class="item-name"><a class="trait" href="trait.Loopback.html" title="trait tower_lsp::Loopback">Loopback</a></div><div class="desc docblock-short">Trait implemented by client loopback sockets.</div></li></ul><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="attr" href="attr.async_trait.html" title="attr tower_lsp::async_trait">async_trait</a></div><div class="desc docblock-short">A re-export of <a href="https://docs.rs/async-trait"><code>async-trait</code></a> for convenience.</div></li></ul></section></div></main></body></html>