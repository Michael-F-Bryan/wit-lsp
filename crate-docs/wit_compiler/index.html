<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `wit_compiler` crate."><title>wit_compiler - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-5bc39a1768837dd0.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="wit_compiler" data-themes="" data-resource-suffix="" data-rustdoc-version="1.77.2 (25ef9e3d8 2024-04-09)" data-channel="1.77.2" data-search-js="search-dd67cee4cfa65049.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-4c98445ec4002617.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-48f368f3872407c8.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-04d5337699b92874.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../wit_compiler/index.html">wit_compiler</a><span class="version">0.2.0</span></h2></div><div class="sidebar-elems"><ul class="block">
            <li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li></ul></section></div></nav><div class="sidebar-resizer"></div>
    <main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../wit_compiler/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">wit_compiler</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../src/wit_compiler/lib.rs.html#1-86">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name" id="reexport.ast"><code>pub use <a class="mod" href="../wit_syntax/index.html" title="mod wit_syntax">wit_syntax</a> as ast;</code></div></li></ul><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="access/index.html" title="mod wit_compiler::access">access</a></div><div class="desc docblock-short">Indirection used when navigating the <a href="hir/index.html" title="mod wit_compiler::hir"><code>crate::hir</code></a> and
<a href="queries/metadata/index.html" title="mod wit_compiler::queries::metadata"><code>crate::queries::metadata</code></a>.</div></li><li><div class="item-name"><a class="mod" href="diagnostics/index.html" title="mod wit_compiler::diagnostics">diagnostics</a></div><div class="desc docblock-short">Errors and user-facing messages that may be generated as the result of
analysis.</div></li><li><div class="item-name"><a class="mod" href="hir/index.html" title="mod wit_compiler::hir">hir</a></div><div class="desc docblock-short">The high-level intermediate representation.</div></li><li><div class="item-name"><a class="mod" href="queries/index.html" title="mod wit_compiler::queries">queries</a></div><div class="desc docblock-short">Various operations used by the compiler and language server.</div></li><li><div class="item-name"><a class="mod" href="traverse/index.html" title="mod wit_compiler::traverse">traverse</a></div><div class="desc docblock-short">Utilities for traversing a <a href="../tree_sitter/struct.Tree.html" title="struct tree_sitter::Tree"><code>tree_sitter::Tree</code></a>.</div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.Compiler.html" title="struct wit_compiler::Compiler">Compiler</a></div><div class="desc docblock-short">The WAI language’s compiler frontend.</div></li><li><div class="item-name"><a class="struct" href="struct.Jar.html" title="struct wit_compiler::Jar">Jar</a></div><div class="desc docblock-short">A <a href="../salsa_2022/jar/trait.Jar.html" title="trait salsa_2022::jar::Jar"><code>salsa::jar::Jar</code></a> containing all queries and tracked types used by this
crate.</div></li><li><div class="item-name"><a class="struct" href="struct.Text.html" title="struct wit_compiler::Text">Text</a></div><div class="desc docblock-short">A cheaply cloneable, immutable string.</div></li><li><div class="item-name"><a class="struct" href="struct.Tree.html" title="struct wit_compiler::Tree">Tree</a></div><div class="desc docblock-short">A wrapper around <a href="../tree_sitter/struct.Tree.html" title="struct tree_sitter::Tree"><code>tree_sitter::Tree</code></a> that is comparable.</div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.Db.html" title="trait wit_compiler::Db">Db</a></div><div class="desc docblock-short">The trait that all <a href="../salsa_2022/database/trait.Database.html" title="trait salsa_2022::database::Database"><code>salsa::Database</code></a> implementations must implement to
be used with this crate’s <a href="queries/index.html" title="mod wit_compiler::queries"><code>queries</code></a>.</div></li></ul></section></div></main></body></html>