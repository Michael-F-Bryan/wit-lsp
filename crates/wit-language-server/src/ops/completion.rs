use tower_lsp::lsp_types::{CompletionItem, CompletionItemKind};
use wit_compiler::{
    access::ScopeIndex,
    ast::{self, AstNode},
    queries::{SourceFile, Workspace},
    Db,
};

pub fn complete(
    db: &dyn Db,
    _ws: Workspace,
    file: SourceFile,
    point: tree_sitter::Point,
) -> Vec<Completion> {
    let mut completions = Vec::new();

    // First, add all known keywords
    let keywords = wit_compiler::ast::KEYWORDS.iter().map(|kw| Completion {
        text: kw.to_string(),
        kind: CompletionKind::Keyword,
    });
    completions.extend(keywords);

    // Now, we do the hard job of figuring out which identifiers are in scope.
    // FIXME: This doesn't take imported items into account
    let items = wit_compiler::queries::file_items(db, file);
    if let Some(index) = items.enclosing_item(db, point) {
        let types = match index {
            ScopeIndex::World(index) => items.get_world(db, index).items(db),
            ScopeIndex::Interface(index) => items.get_interface(db, index).items(db),
        };

        completions.extend(types.names().map(|name| Completion {
            text: name.to_string(),
            kind: CompletionKind::TypeName,
        }));
    }

    let ast = wit_compiler::queries::parse(db, file);
    if let Some(ident) = ast
        .tree(db)
        .ancestors(point)
        .find_map(ast::Identifier::cast)
    {
        let src = file.contents(db);
        let ident = ident.value(src);
        // The user has started writing an identifier, so limit the completions
        // to whatever might match what they've written.
        completions.retain(|c| c.text.starts_with(ident));
    }

    completions
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Completion {
    pub text: String,
    pub kind: CompletionKind,
}

impl From<Completion> for CompletionItem {
    fn from(value: Completion) -> Self {
        CompletionItem {
            label: value.text,
            kind: Some(value.kind.into()),
            ..Default::default()
        }
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum CompletionKind {
    TypeName,
    Interface,
    World,
    Keyword,
}

impl From<CompletionKind> for CompletionItemKind {
    fn from(value: CompletionKind) -> Self {
        match value {
            CompletionKind::TypeName => CompletionItemKind::STRUCT,
            CompletionKind::Interface => CompletionItemKind::INTERFACE,
            CompletionKind::World => CompletionItemKind::MODULE,
            CompletionKind::Keyword => CompletionItemKind::KEYWORD,
        }
    }
}

#[cfg(test)]
mod tests {
    use std::borrow::Cow;

    use im::OrdMap;
    use wit_compiler::{queries::FilePath, Compiler, Tree};

    use super::*;

    fn find_completion_location<'a>(
        tree: &Tree,
        src: &'a str,
    ) -> OrdMap<Cow<'a, str>, tree_sitter::Point> {
        tree.iter()
            .filter_map(wit_compiler::ast::BlockComment::cast)
            .map(|comment| (comment.text(src), comment.range().start_point))
            .collect()
    }

    #[test]
    fn completions_in_different_contexts() {
        let src = r#"
            interface foo {
                /* first */

                type x = /* second */ u32;
            }

            world bar {
                export /* third */;
            }
        "#;
        let db = Compiler::default();
        let path = FilePath::new(&db, "complete_builtins.wit".into());
        let file = SourceFile::new(&db, path, src.into());
        let ws = Workspace::new(&db, OrdMap::unit(path, file));
        let ast = wit_compiler::queries::parse(&db, file);
        let locations = find_completion_location(ast.tree(&db), src);

        let result: OrdMap<String, Vec<Completion>> = locations
            .into_iter()
            .map(|(k, p)| (k.to_string(), complete(&db, ws, file, p)))
            .collect();

        insta::assert_debug_snapshot!(result);
    }
}
