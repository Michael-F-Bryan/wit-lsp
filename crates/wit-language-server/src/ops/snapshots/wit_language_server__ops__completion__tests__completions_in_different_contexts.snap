---
source: crates/wit-language-server/src/ops/completion.rs
expression: result
---
{
    "first": [
        Completion {
            text: "as",
            kind: Keyword,
        },
        Completion {
            text: "bool",
            kind: Keyword,
        },
        Completion {
            text: "borrow",
            kind: Keyword,
        },
        Completion {
            text: "char",
            kind: Keyword,
        },
        Completion {
            text: "constructor",
            kind: Keyword,
        },
        Completion {
            text: "enum",
            kind: Keyword,
        },
        Completion {
            text: "export",
            kind: Keyword,
        },
        Completion {
            text: "flags",
            kind: Keyword,
        },
        Completion {
            text: "float32",
            kind: Keyword,
        },
        Completion {
            text: "float64",
            kind: Keyword,
        },
        Completion {
            text: "func",
            kind: Keyword,
        },
        Completion {
            text: "import",
            kind: Keyword,
        },
        Completion {
            text: "include",
            kind: Keyword,
        },
        Completion {
            text: "interface",
            kind: Keyword,
        },
        Completion {
            text: "list",
            kind: Keyword,
        },
        Completion {
            text: "option",
            kind: Keyword,
        },
        Completion {
            text: "own",
            kind: Keyword,
        },
        Completion {
            text: "package",
            kind: Keyword,
        },
        Completion {
            text: "record",
            kind: Keyword,
        },
        Completion {
            text: "resource",
            kind: Keyword,
        },
        Completion {
            text: "result",
            kind: Keyword,
        },
        Completion {
            text: "s16",
            kind: Keyword,
        },
        Completion {
            text: "s32",
            kind: Keyword,
        },
        Completion {
            text: "s64",
            kind: Keyword,
        },
        Completion {
            text: "s8",
            kind: Keyword,
        },
        Completion {
            text: "static",
            kind: Keyword,
        },
        Completion {
            text: "string",
            kind: Keyword,
        },
        Completion {
            text: "tuple",
            kind: Keyword,
        },
        Completion {
            text: "type",
            kind: Keyword,
        },
        Completion {
            text: "u16",
            kind: Keyword,
        },
        Completion {
            text: "u32",
            kind: Keyword,
        },
        Completion {
            text: "u64",
            kind: Keyword,
        },
        Completion {
            text: "u8",
            kind: Keyword,
        },
        Completion {
            text: "use",
            kind: Keyword,
        },
        Completion {
            text: "variant",
            kind: Keyword,
        },
        Completion {
            text: "with",
            kind: Keyword,
        },
        Completion {
            text: "world",
            kind: Keyword,
        },
        Completion {
            text: "x",
            kind: TypeName,
        },
    ],
    "second": [
        Completion {
            text: "as",
            kind: Keyword,
        },
        Completion {
            text: "bool",
            kind: Keyword,
        },
        Completion {
            text: "borrow",
            kind: Keyword,
        },
        Completion {
            text: "char",
            kind: Keyword,
        },
        Completion {
            text: "constructor",
            kind: Keyword,
        },
        Completion {
            text: "enum",
            kind: Keyword,
        },
        Completion {
            text: "export",
            kind: Keyword,
        },
        Completion {
            text: "flags",
            kind: Keyword,
        },
        Completion {
            text: "float32",
            kind: Keyword,
        },
        Completion {
            text: "float64",
            kind: Keyword,
        },
        Completion {
            text: "func",
            kind: Keyword,
        },
        Completion {
            text: "import",
            kind: Keyword,
        },
        Completion {
            text: "include",
            kind: Keyword,
        },
        Completion {
            text: "interface",
            kind: Keyword,
        },
        Completion {
            text: "list",
            kind: Keyword,
        },
        Completion {
            text: "option",
            kind: Keyword,
        },
        Completion {
            text: "own",
            kind: Keyword,
        },
        Completion {
            text: "package",
            kind: Keyword,
        },
        Completion {
            text: "record",
            kind: Keyword,
        },
        Completion {
            text: "resource",
            kind: Keyword,
        },
        Completion {
            text: "result",
            kind: Keyword,
        },
        Completion {
            text: "s16",
            kind: Keyword,
        },
        Completion {
            text: "s32",
            kind: Keyword,
        },
        Completion {
            text: "s64",
            kind: Keyword,
        },
        Completion {
            text: "s8",
            kind: Keyword,
        },
        Completion {
            text: "static",
            kind: Keyword,
        },
        Completion {
            text: "string",
            kind: Keyword,
        },
        Completion {
            text: "tuple",
            kind: Keyword,
        },
        Completion {
            text: "type",
            kind: Keyword,
        },
        Completion {
            text: "u16",
            kind: Keyword,
        },
        Completion {
            text: "u32",
            kind: Keyword,
        },
        Completion {
            text: "u64",
            kind: Keyword,
        },
        Completion {
            text: "u8",
            kind: Keyword,
        },
        Completion {
            text: "use",
            kind: Keyword,
        },
        Completion {
            text: "variant",
            kind: Keyword,
        },
        Completion {
            text: "with",
            kind: Keyword,
        },
        Completion {
            text: "world",
            kind: Keyword,
        },
        Completion {
            text: "x",
            kind: TypeName,
        },
    ],
    "third": [
        Completion {
            text: "as",
            kind: Keyword,
        },
        Completion {
            text: "bool",
            kind: Keyword,
        },
        Completion {
            text: "borrow",
            kind: Keyword,
        },
        Completion {
            text: "char",
            kind: Keyword,
        },
        Completion {
            text: "constructor",
            kind: Keyword,
        },
        Completion {
            text: "enum",
            kind: Keyword,
        },
        Completion {
            text: "export",
            kind: Keyword,
        },
        Completion {
            text: "flags",
            kind: Keyword,
        },
        Completion {
            text: "float32",
            kind: Keyword,
        },
        Completion {
            text: "float64",
            kind: Keyword,
        },
        Completion {
            text: "func",
            kind: Keyword,
        },
        Completion {
            text: "import",
            kind: Keyword,
        },
        Completion {
            text: "include",
            kind: Keyword,
        },
        Completion {
            text: "interface",
            kind: Keyword,
        },
        Completion {
            text: "list",
            kind: Keyword,
        },
        Completion {
            text: "option",
            kind: Keyword,
        },
        Completion {
            text: "own",
            kind: Keyword,
        },
        Completion {
            text: "package",
            kind: Keyword,
        },
        Completion {
            text: "record",
            kind: Keyword,
        },
        Completion {
            text: "resource",
            kind: Keyword,
        },
        Completion {
            text: "result",
            kind: Keyword,
        },
        Completion {
            text: "s16",
            kind: Keyword,
        },
        Completion {
            text: "s32",
            kind: Keyword,
        },
        Completion {
            text: "s64",
            kind: Keyword,
        },
        Completion {
            text: "s8",
            kind: Keyword,
        },
        Completion {
            text: "static",
            kind: Keyword,
        },
        Completion {
            text: "string",
            kind: Keyword,
        },
        Completion {
            text: "tuple",
            kind: Keyword,
        },
        Completion {
            text: "type",
            kind: Keyword,
        },
        Completion {
            text: "u16",
            kind: Keyword,
        },
        Completion {
            text: "u32",
            kind: Keyword,
        },
        Completion {
            text: "u64",
            kind: Keyword,
        },
        Completion {
            text: "u8",
            kind: Keyword,
        },
        Completion {
            text: "use",
            kind: Keyword,
        },
        Completion {
            text: "variant",
            kind: Keyword,
        },
        Completion {
            text: "with",
            kind: Keyword,
        },
        Completion {
            text: "world",
            kind: Keyword,
        },
    ],
}
